src/websocket-events/websocket-events.base.ts

```diff
@@ -0,0 +1,97 @@
+import * as http from 'http';
+import { Logger } from '@nestjs/common';
+import {
+  OnGatewayConnection,
+  OnGatewayDisconnect,
+  OnGatewayInit,
+  WebSocketServer,
+} from '@nestjs/websockets';
+import { nanoid } from 'nanoid';
+import WebSocket, { Server } from 'ws';
+
+const PING_INTERVAL = 5000;
+
+function checkApiKey(_request: http.IncomingMessage) {
+  // TODO: implement
+  return false;
+}
+
+export interface WebSocketEx extends WebSocket {
+  isAlive: boolean;
+  id: string;
+  request?: http.IncomingMessage;
+}
+
+/**
+ * Message format expected by @SubscribeMessage()
+ *
+ * While technically the websocket messages (in both directions) could
+ * use any arbitrary format, this format is simple and has the advantage
+ * of playing nice with NestJS decorators for incoming events.
+ */
+export interface WebSocketMessage {
+  event: string;
+  data: any;
+}
+
+/**
+ * Base class for websocket gateways.
+ *
+ * To create a new gateway, subclass and decorate your child, e.g.:
+ * @WebSocketGateway({ path: '/api/ws' })
+ */
+export abstract class WebSocketEventsBase
+  implements OnGatewayInit, OnGatewayConnection, OnGatewayDisconnect
+{
+  @WebSocketServer() server: Server;
+
+  constructor(protected readonly logger: Logger, private requireAuth = false) {}
+
+  afterInit(server: Server) {
+    const interval = setInterval(() => this.ping(), PING_INTERVAL);
+    server.on('connection', (client: WebSocketEx, req) => {
+      client.id = nanoid();
+      client.isAlive = true;
+      client.request = req;
+      if (this.requireAuth && !checkApiKey(req)) {
+        this.logger.log(`WebSocket ${client.id}: unauthorized`);
+        client.close(1008, 'Unauthorized');
+      }
+    });
+    server.on('close', () => clearInterval(interval));
+  }
+
+  handleConnection(client: WebSocketEx) {
+    this.logger.log(`WebSocket ${client.id}: connected`);
+
+    client.on('pong', () => {
+      client.isAlive = true;
+    });
+    client.on('error', err => {
+      this.logger.log(`WebSocket ${client.id}: error: ${err}`);
+    });
+  }
+
+  handleDisconnect(client: WebSocketEx) {
+    this.logger.log(`WebSocket ${client.id}: disconnected`);
+  }
+
+  ping() {
+    this.server?.clients.forEach(ws => {
+      const client = ws as WebSocketEx;
+      if (client.isAlive === false) {
+        this.logger.log(`WebSocket ${client.id}: connection lost`);
+        return client.terminate();
+      }
+      client.isAlive = false;
+      client.ping();
+    });
+  }
+
+  broadcast(event: string, data: any = null) {
+    const payload = JSON.stringify(<WebSocketMessage>{ event, data });
+    const clients = this.server.clients;
+    clients.forEach(client => client.send(payload));
```

Comment:
The contract between FireFly and Tokens cannot be broadcast, instead it needs to be either:
1. One client connection gets every event, until it disconnects and then all events start going to the next
2. Each broadcast is workload managed between available clients

I note that (2) requires handling of `ack` between FireFly and Tokens before the next message can be sent, whereas I think (1) can rely on the fact that a single Go routine on the FireFly side is processing the events sequentially.

Reply:
Right, I meant to circle back on this item and hadn't yet. Thanks for catching.

Reply:
Implemented (1) with the latest changes.